cli/src/main/java/hudson/cli/PlainCLIProtocol.java

```diff
@@ -0,0 +1,344 @@
+                        }
+                        Op op = Op.values()[b];
+                        long start = cis.getByteCount();
+                        LOGGER.log(Level.FINEST, "handling frame with {0} of length {1}", new Object[] {op, framelen});
+                        boolean handled = handle(op, framelen);
+                        if (handled) {
+                            long actuallyRead = cis.getByteCount() - start;
+                            if (actuallyRead != framelen) {
+                                throw new IOException("corrupt stream: expected to read " + framelen + " bytes from " + op + " but read " + actuallyRead);
+                            }
+                        } else {
+                            LOGGER.log(Level.WARNING, "unexpected {0}: {1}", new Object[] {op, HexDump.toHex(flightRecorder.getRecord())});
+                            IOUtils.skipFully(dis, framelen);
+                        }
+                    }
+                } catch (ClosedChannelException x) {
+                    LOGGER.log(Level.FINE, null, x);
+                    handleClose();
+                } catch (IOException x) {
+                    LOGGER.log(Level.WARNING, null, flightRecorder.analyzeCrash(x, "broken stream"));
+                } catch (ReadPendingException x) {
+                    // in case trick in CLIAction does not work
+                    LOGGER.log(Level.FINE, null, x);
+                    handleClose();
+                }
+            }
+
+        }
+
+        protected abstract void handleClose();
+
+        protected abstract boolean handle(Op op, int framelen) throws IOException;
+
+        private void writeOp(Op op) throws IOException {
+            dos.writeByte((byte) op.ordinal());
+        }
+
+        protected final synchronized void send(Op op) throws IOException {
+            dos.writeInt(0);
+            writeOp(op);
+            dos.flush();
+        }
+
+        protected final synchronized void send(Op op, int number) throws IOException {
+            dos.writeInt(4);
+            writeOp(op);
+            dos.writeInt(number);
+            dos.flush();
+        }
+
+        protected final synchronized void send(Op op, byte[] chunk, int off, int len) throws IOException {
+            dos.writeInt(len);
+            writeOp(op);
+            dos.write(chunk, off, len);
+            dos.flush();
+        }
+
+        protected final void send(Op op, byte[] chunk) throws IOException {
+            send(op, chunk, 0, chunk.length);
+        }
+
+        protected final void send(Op op, String text) throws IOException {
+            ByteArrayOutputStream buf = new ByteArrayOutputStream();
+            new DataOutputStream(buf).writeUTF(text);
+            send(op, buf.toByteArray());
+        }
+
+        protected final byte[] readChunk(int framelen) throws IOException {
+            assert Thread.currentThread() instanceof EitherSide.Reader;
+            byte[] buf = new byte[framelen];
```

Comment:
Would be possible. For most commands performance of stdio is not an issue. The exceptions would be `console` (or `build -v`).